engine/runtime-integration-tests/src/test/java/org/enso/compiler/test/passes/resolve/FullyQualifiedNamesTest.java

```diff
@@ -0,0 +1,241 @@
+import org.enso.scala.wrapper.ScalaConversions;
+import org.enso.test.utils.ContextUtils;
+import org.enso.test.utils.ProjectUtils;
+import org.enso.test.utils.SourceModule;
+import org.graalvm.polyglot.Context;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+public final class FullyQualifiedNamesTest {
+
+  @Rule public final TemporaryFolder TMP_DIR = new TemporaryFolder();
+
+  @Test
+  public void libraryNameIsResolved_InExpressionBlock() throws IOException {
+    String mainSrc = """
+        main =
+            local.Proj.My_Module.My_Type
+        """;
+    var projDir = TMP_DIR.newFolder("Proj").toPath();
+    ProjectUtils.createProject("Proj", mainSrc, projDir);
+    try (var ctx = createCtx(projDir)) {
+      compileAllModules(ctx);
+      var modIr = getModuleIr(ctx, "local.Proj.Main");
+      var location = getLastLocationOf(mainSrc, "Proj");
+      var ir = findIrByLocation(modIr, location);
+      assertHasFQNMetadata(
+          ir,
+          FullyQualifiedNames.ResolvedModule.class,
+          meta -> {
+            assertThat(meta.moduleRef().getName().toString(), is("local.Proj.Main"));
+          });
+    }
+  }
+
+  /**
+   * Fully qualified name used in {@link Pattern.Type} is resolved in {@link Patterns} compiler
+   * pass. Moreover, this pass resolves it to {@link ResolvedType} - which means full resolution.
+   */
+  @Test
+  public void libNameIsResolved_InCaseBranch() throws Exception {
+    var myMod = srcModule("My_Module", """
+        type My_Type
+        """);
+    var mainMod =
+        srcModule(
+            "Main",
+            """
+        import project.My_Module.My_Type
+
+        main =
+            x = 42
+            case x of
+                _ : local.Proj.My_Module.My_Type -> 21  # Case.Branch
+                _ -> 22
+        """);
+    var projDir = TMP_DIR.newFolder("Proj").toPath();
+    ProjectUtils.createProject("Proj", Set.of(myMod, mainMod), projDir);
+    try (var ctx = createCtx(projDir)) {
+      compileAllModules(ctx);
+      var modIr = getModuleIr(ctx, "local.Proj.Main");
+      var caseBranch =
+          findIR(
+              modIr,
+              Case.Branch.class,
+              branch -> {
+                if (branch.expression() instanceof Literal.Number num) {
+                  return num.value().equals("21");
+                }
+                return false;
+              });
+      assertThat(caseBranch, is(notNullValue()));
+      var patternType = (Pattern.Type) caseBranch.pattern();
+      var tpeName = patternType.tpe();
+      assertHasMetadata(
+          tpeName,
+          Patterns$.MODULE$,
+          BindingsMap.Resolution.class,
+          resolution -> {
+            assertThat(
+                "Resolution target: " + resolution.target() + " should be ResolvedType",
+                resolution.target() instanceof ResolvedType,
+                is(true));
+            assertThat(
```

Comment:
[Patterns](https://github.com/enso-org/enso/blob/b5f110617e1b96c118c6483bddaf982c484c8945/engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/resolve/Patterns.scala) IR pass resolves `local.Proj.My_Module.My_Type` to the correct type. So why [FullyQualifiedNames](https://github.com/enso-org/enso/blob/9cd5ca30bf1a7799900836a6dc671272d8b09d27/engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/resolve/FullyQualifiedNames.scala) does not resolve that as well? Name resolution in compiler seems to be all over the place. Let's try to create some NameResolver that will be used by both `Pattern` and `FullyQualifiedNames`.